etcd/src/main/scala/io/buoyant/etcd/Key.scala

```diff
@@ -0,0 +1,278 @@
+   * provided `index` value.
+   *
+   * If `prevValue` is specified, the current node must have the
+   * provided value.
+   *
+   * If `prevExist` is specified, the node must already exist.
+   */
+  def compareAndSwap(
+    value: Buf,
+    prevIndex: Option[Long] = None,
+    prevValue: Option[Buf] = None,
+    prevExist: Boolean = true
+  ): Future[NodeOp] = {
+    require(prevIndex.isDefined || prevValue.isDefined || !prevExist)
+
+    var params = Seq.empty[(String, String)]
+    for (i <- prevIndex) {
+      params = params :+ "prevIndex" -> i.toString
+    }
+    for (Buf.Utf8(v) <- prevValue) {
+      params = params :+ "prevValue" -> v
+    }
+    if (!prevExist) {
+      params = params :+ "prevExist" -> "false"
+    }
+    val Buf.Utf8(v) = value
+    params = params :+ "value" -> v
+
+    val req = mkReq(uriPath, Method.Put, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /**
+   * Delete a node.
+   *
+   * If `dir` is not true and the key is a directory, this will
+   * operation fail.
+   *
+   * If `dir` and `recursive` are true, the entire tree is deleted.
+   */
+  def delete(
+    dir: Boolean = false,
+    recursive: Boolean = false
+  ): Future[NodeOp] = {
+    var params = Seq.empty[(String, String)]
+    params +:= "foo" -> "bar"
+    if (dir) {
+      params = params :+ ("dir" -> "true")
+      if (recursive) {
+        params = params :+ ("recursive" -> "true")
+      }
+    }
+    val req = mkReq(uriPath, Method.Delete, params)
+    client(req).flatMap { rsp => Future.const(NodeOp.mk(req, rsp, key, params)) }
+  }
+
+  /** Find the greatest index referenced in a tree of nodes. */
+  private[this] def getIndex(node: Node): Long = node match {
+    case Node.Data(_, idx, _, _, _) => idx
+    case Node.Dir(_, idx, _, _, nodes) => nodes.foldLeft(idx)(_ max getIndex(_))
+  }
+
+  /**
+   * An Event that reflects
+   *
+   * If `recursive` is true, the key's subtree is observed for
+   * changes.
+   *
+   * When an unexpected error is encountered communicating with the
+   * API, the failure is published on the Event and the `backoff`
+   * stream is used to compute the time to wait before retrying. If
+   * the `backoff` stream is exhausted or a fatal error is
+   * encountered, it is reported and polling stops.
+   *
+   * The Event is not reference-counted, so each observer initiates
+   * its own polling loop. This ensures, for instance, that the
+   * initial state of a tree is reported properly.
+   */
+  def events(
+    recursive: Boolean = false,
+    backoff: Stream[Duration] = Stream.empty
+  ): Event[Try[NodeOp]] = new Event[Try[NodeOp]] {
+    private[this] val origBackoff = backoff
+
+    def register(witness: Witness[Try[NodeOp]]) = {
+      @volatile var closed = false
+      @volatile var currentOp: Future[NodeOp] = Future.never
+
+      def loop(idx: Option[Long], backoff: Stream[Duration]): Unit =
```

Comment:
TIOLI: This could return `Future[Unit]` and use `flatMap` instead of `respond`.  That way you would't need `currentOp` and could just raise on the total `Future[Unit]` returned from the original invocation of `loop`.